execution/function/function.go

```diff
@@ -1,247 +1,474 @@
+	series      []labels.Labels
+	once        sync.Once
+	vectorIndex int
+	nextOps     []model.VectorOperator
+
+	call         FunctionCall
+	scalarPoints [][]float64
+}
+
+func NewfunctionOperator(funcExpr *parser.Call, call FunctionCall, nextOps []model.VectorOperator, stepsBatch int) (model.VectorOperator, error) {
+	scalarPoints := make([][]float64, stepsBatch)
+	for i := 0; i < stepsBatch; i++ {
+		scalarPoints[i] = make([]float64, len(nextOps)-1)
+	}
+	f := &functionOperator{
+		nextOps:      nextOps,
+		call:         call,
+		funcExpr:     funcExpr,
+		vectorIndex:  0,
+		scalarPoints: scalarPoints,
+	}
+
+	for i := range funcExpr.Args {
+		if funcExpr.Args[i].Type() == parser.ValueTypeVector {
+			f.vectorIndex = i
+			break
+		}
+	}
+
+	// Check selector type.
+	// TODO(saswatamcode): Add support for string and matrix.
+	switch funcExpr.Args[f.vectorIndex].Type() {
+	case parser.ValueTypeVector, parser.ValueTypeScalar:
+		return f, nil
+	default:
+		return nil, errors.Wrapf(parse.ErrNotImplemented, "got %s:", funcExpr.String())
+	}
+}
+
+func (o *functionOperator) Explain() (me string, next []model.VectorOperator) {
+	return fmt.Sprintf("[*functionOperator] %v(%v)", o.funcExpr.Func.Name, o.funcExpr.Args), o.nextOps
+}
+
+func (o *functionOperator) Series(ctx context.Context) ([]labels.Labels, error) {
+	if err := o.loadSeries(ctx); err != nil {
+		return nil, err
+	}
+
+	return o.series, nil
+}
+
+func (o *functionOperator) GetPool() *model.VectorPool {
+	return o.nextOps[o.vectorIndex].GetPool()
+}
+
+func (o *functionOperator) Next(ctx context.Context) ([]model.StepVector, error) {
+	if err := o.loadSeries(ctx); err != nil {
+		return nil, err
+	}
+
+	// Process non-variadic single/multi-arg instant vector and scalar input functions.
+	// Call next on vector input.
+	vectors, err := o.nextOps[o.vectorIndex].Next(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(vectors) == 0 {
+		return nil, nil
+	}
+
+	scalarIndex := 0
+	for i := range o.nextOps {
+		if i == o.vectorIndex {
+			continue
+		}
+
+		scalarVectors, err := o.nextOps[i].Next(ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		for batchIndex := range vectors {
+			val := math.NaN()
+			if len(scalarVectors) > 0 && len(scalarVectors[batchIndex].Samples) > 0 {
+				val = scalarVectors[batchIndex].Samples[0]
+				o.nextOps[i].GetPool().PutStepVector(scalarVectors[batchIndex])
+			}
+			o.scalarPoints[batchIndex][scalarIndex] = val
+		}
+		o.nextOps[i].GetPool().PutVectors(scalarVectors)
+		scalarIndex++
+	}
+
+	for batchIndex, vector := range vectors {
+		// scalar() depends on number of samples per vector and returns NaN if len(samples) != 1.
+		// So need to handle this separately here, instead of going via call which is per point.
+		if o.funcExpr.Func.Name == "scalar" {
+			if len(vector.Samples) <= 1 {
+				continue
+			}
+
+			vector.Samples = vector.Samples[:1]
+			vector.SampleIDs = vector.SampleIDs[:1]
+			vector.Samples[0] = math.NaN()
```

Comment:
Why do we set this to nan?

Reply:
`scalar()` will return NaN if the given input vector has more than one sample as mentioned here: https://prometheus.io/docs/prometheus/latest/querying/functions/#scalar.

Reply:
Thanks for the explanation.